linked_list/linked_list.py

```diff
@@ -84,42 +131,70 @@ def visit(self):
         
         print(", ".join(helper_list))
 
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
+        if self.head == None:
+            return
+        if self.head.next is None:
+            return self.head
+        prev = None
+        current = self.head
+        while current is not None:
+            next = current.next
+            current.next = prev
+            prev = current
+            current = next
+        self.head = prev
   
-    ## Advanced/ Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        if self.head == None:
+            return
+        if self.head.next is None:
+            return self.head
+        current = self.head
+        count = 0
+        n = self.length()//2
+        while current is not None:
+            if count == n:
+                return current.value
+            count += 1
+            current = current.next
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):

```

Comment:
👍  Good use of prior methods and some arithmetic.  However why not use `get_at_index`?